file_formats/cmt_format.ml

```diff
@@ -197,12 +213,169 @@ let clear_env binary_annots =
 
   else binary_annots
 
+let iter_on_usages ~index =
+  let f ~namespace env path lid =
+    let not_ghost { Location.loc = { loc_ghost; _ }; _ } = not loc_ghost in
+    if not_ghost lid then
+      try
+        let shape = Env.shape_of_path ~namespace env path in
+        let shape = Local_reduce.weak_reduce env shape in
+        if not (Shape.is_closed shape) then
+          index := (Unresolved shape, lid) :: !index
+        else Option.iter
+          (fun uid -> index := (Resolved uid, lid) :: !index)
+          shape.Shape.uid
```

Comment:
Am I correct in thinking we should always have an uid here?

Reply:
If the reduction was stuck and the shape is approximation the uid might be missing.
If the reduction finished then there should always be a Uid. If that's not the case this is most probably a bug. Your comment allow me to find one such missing uid for local module. I fixed that in b54012b .
[6bbacf9](https://github.com/ocaml/ocaml/pull/12508/commits/6bbacf916c9e97f563d5fa4fdc929f4f4048b6fd) add more constructors to the index list to report missing uids and approximate shapes.